ToolsPanel: Standardize input control label margin

aaronrobertshaw opened the pull request:
Related: https://github.com/WordPress/gutenberg/pull/36334

## Description

This PR standardizes the line-height and margin of `InputControl` labels within the `ToolsPanel` to improve consistency.

See: https://github.com/WordPress/gutenberg/pull/36334#issuecomment-965806487

## How has this been tested?

1. Open editor and add text to a paragraph
2. With paragraph block selected, toggle on display of both line height and letter spacing controls from the typography panel
3. Ensure the labels and spacing looks consistent

## Screenshots <!-- if applicable -->

| Before | After |
|---|---|
| <img width="278" alt="Screen Shot 2021-11-11 at 11 39 16 am" src="https://user-images.githubusercontent.com/60436221/141223497-daf100e1-3fd2-4526-bec5-c55a1bfdccb6.png"> | <img width="281" alt="Screen Shot 2021-11-11 at 11 38 55 am" src="https://user-images.githubusercontent.com/60436221/141223519-6c249873-c099-43dc-b38b-d03f37601d9f.png"> |

## Types of changes
Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Standardize input control label margin within ToolsPanel

aaronrobertshaw commented on packages/components/src/tools-panel/styles.ts:
```diff
@@ -107,6 +108,15 @@ export const ToolsPanelItem = css`
 	}
+
+	/* Standardize InputControl labels with others inside ToolsPanel. */
+	&& ${ LabelWrapper } {
```
I ended up using `LabelWrapper` from the input control styles here as I couldn't get `BaseLabel` to work for me. Open to suggestions to improve this.

ramonjd reviewed (approved):
Very nice. Looks 👍 to me. Thanks for the elegant solution.

| Before        | After           |
| :---: | :---: |
| <img width="281" alt="Screen Shot 2021-11-11 at 1 33 58 pm" src="https://user-images.githubusercontent.com/6458278/141227369-3fbbb136-78e2-4dfe-b2b4-30b7cd53efb7.png"> | <img width="281" alt="Screen Shot 2021-11-11 at 1 36 40 pm" src="https://user-images.githubusercontent.com/6458278/141227358-81a40a7c-65e4-4ddf-9537-674e4ef4c941.png"> |

ciampo reviewed (commented):
I inspected the control components and the corresponding labels:

- The "Appearance" control is based on the `CustomSelectControl` component, which internally [uses a "vanilla" `<label>`](https://github.com/WordPress/gutenberg/blob/e88eb9839306f8a524e95c88213f947cd5fe1622/packages/components/src/custom-select-control/index.js#L120-L126)
- The "Letter spacing" control is based on the `UnitControl` component, which builds on top of the `InputControl` component ([including its label](https://github.com/WordPress/gutenberg/blob/92662feeb220336fb28f8b79869cb251a7db4e27/packages/components/src/input-control/label.tsx))

Instead of solving this problem specifically to `ToolsPanel`, I think that a better approach here would be to align the label in`CustomSelectControl` to have the same styling (including line height, margins and paddings) to the one of `InputControl`.

Not sure if it's the best solution here, but maybe `CustomSelectControl` could use the `BaseLabel` directly from `InputControl`?  (Or, in general, have `CustomSelectControl` be based off `InputControl` as much as possible) ?

---

Also a side note — should we update [this Storybook example](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-toolspanel--typography-panel) to allow single-columns controls? It would be another good way to test for the changes in this PR in an isolated environment

aaronrobertshaw commented:
> I inspected the control components and the corresponding labels:

I think there is another type of control missed here, the `BaseControl`.

Just within the Typography panel alone we have ~three~ four sources for different labels:

- `CustomSelectControl` --> HTML `label` 
- `InputControl` --> `BaseLabel`
- `BaseControl` --> `StyledLabel`
- Text Decoration/Transform controls --> HTML `legend`

I'd be inclined to believe the `BaseControl` should be the starting point. Plus it actually provides a margin so the label isn't cramped against the input.

> Instead of solving this problem specifically to ToolsPanel, I think that a better approach here would be to align the label inCustomSelectControl to have the same styling (including line height, margins and paddings) to the one of InputControl.

The `InputControl` appears to be the odd one out here in terms of styling. 

I wanted to avoid making any breaking changes (even if only styling) outside of the `ToolsPanel` given these are all likely used elsewhere. It feels like it would be better to bring all of these inline in a separate PR once all the required components and those built off them are switch to Emotion and the context system.

> Also a side note — should we update this Storybook example to allow single-columns controls? It would be another good way to test for the changes in this PR in an isolated environment

My understanding was that Storybook example was a playground for developing the new controls for the Typography panel. I'm not sure it makes sense to try and use it to text/develop the current controls at the same time. A small problem with relying on the Storybook and an isolated environment is we still have to make it line up neatly outside that sterile environment and in the real editors.

ciampo commented:
> I think there is another type of control missed here, the `BaseControl` [...] Just within the Typography panel alone we have three four sources for different labels

Thank you for the sleuthing and the detailed explanation!

> I'd be inclined to believe the `BaseControl` should be the starting point

Refactoring these components to use the same base control/label sounds definitely like the sensible thing to do

> I wanted to avoid making any breaking changes (even if only styling) outside of the ToolsPanel given these are all likely used elsewhere. It feels like it would be better to bring all of these inline in a separate PR once all the required components and those built off them are switch to Emotion and the context system.

This is definitely a sensible point — **I'd be ok with this PR focusing on a "hotfix" specific for `ToolsPanel`.**

But at the same time it'd be great if we could **follow up and look into the discrepancies across "control" components, come up with a good understanding/dependency graph, and work towards a normalization** (in the direction that you suggested above) **soon**.

> My understanding was that Storybook example was a playground for developing the new controls for the Typography panel. I'm not sure it makes sense to try and use it to text/develop the current controls at the same time. A small problem with relying on the Storybook and an isolated environment is we still have to make it line up neatly outside that sterile environment and in the real editors.

I agree that relying on Storybook alone is not a good idea, since it is a "sterile" environment, and that ultimately things need to look good in the real editors. At the same time, I also believe that it's important to **use Storybook as the first "test stage" for our components**.

As an example, I tried to to update Storybook so that the Line Height and Letter Spacing controls are next to each other, and noticed a misalignment

![image](https://user-images.githubusercontent.com/1083581/141451195-f1acd41e-97cd-43f5-abe8-05157eb8ed94.png)

<details>
<summary>Code used to update Storybook</summary>

```diff
diff --git a/packages/components/src/tools-panel/stories/typography-panel.js b/packages/components/src/tools-panel/stories/typography-panel.js
index c57ac4c19e..afd8646788 100644
--- a/packages/components/src/tools-panel/stories/typography-panel.js
+++ b/packages/components/src/tools-panel/stories/typography-panel.js
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import { css } from '@emotion/react';
+
 /**
  * WordPress dependencies
  */
@@ -12,6 +17,7 @@ import { useState } from '@wordpress/element';
 /**
  * Internal dependencies
  */
+import { useCx } from '../../utils/hooks/use-cx';
 import FontSizePicker from '../../font-size-picker';
 import Panel from '../../panel';
 import {
@@ -97,6 +103,10 @@ export const TypographyPanel = () => {
 	const [ letterSpacing, setLetterSpacing ] = useState();
 	const [ textTransform, setTextTransform ] = useState();
 
+	const cx = useCx();
+
+	const singleColumnClassName = cx( css( { gridColumn: 'span 1' } ) );
+
 	return (
 		<>
 			<div style={ { maxWidth: '280px' } }>
@@ -157,6 +167,7 @@ export const TypographyPanel = () => {
 							label="Line Height"
 							onDeselect={ () => setLineHeight( undefined ) }
 							isShownByDefault={ true }
+							className={ singleColumnClassName }
 						>
 							<LineHeightControl
 								value={ lineHeight }
@@ -169,6 +180,7 @@ export const TypographyPanel = () => {
 							label="Letter Spacing"
 							onDeselect={ () => setLetterSpacing( undefined ) }
 							isShownByDefault={ true }
+							className={ singleColumnClassName }
 						>
 							<LetterSpacingControl
 								value={ letterSpacing }
```

</details>

What I'm suggesting is **a component-first approach** (as opposed to an editor-first approach). In order to do that:

1. We should make sure that a component is displayed correctly in Storybook first. This would guarantee that the component works as expected "out o the box"
2. We should then test it in the real code editor, and see if there are any discrepancies. In case there are, we should investigate the issue, and always try to fix it at the source (the component), and restart the process from point 1. When that's not possible, we would need to resort to applying a fix in the editor code

ciampo commented on packages/components/src/tools-panel/styles.ts:
```diff
@@ -107,6 +108,15 @@ export const ToolsPanelItem = css`
 		}
 	}
+
+	/* Standardize InputControl labels with others inside ToolsPanel. */
```
We should add a comment here about this being a temporary fix — with the proper fix being the normalization of the different control components

aaronrobertshaw committed:
Ensure BaseControl label line-height in ToolsPanel

This was in response to issue raised about misalignment of labels only in the Storybook.

aaronrobertshaw commented:
> As an example, I tried to to update Storybook so that the Line Height and Letter Spacing controls are next to each other, and noticed a misalignment

This appears to be due to the lack of a standard line-height applied to labels in the Storybook.

I've added a new rule that ensures that standardized line-height for `BaseControl` labels within the tool panel and Storybook.

<img width="309" alt="Screen Shot 2021-11-16 at 6 12 18 pm" src="https://user-images.githubusercontent.com/60436221/141948602-1bb5cc59-04c0-4bc9-8315-f8146c9a10ea.png">

You'll notice that:
- Line height and letter spacing labels align
- Font size component effectively has it's label heightened due to the toggle icon
- The proposed letter-case control uses `ToggleGroupControl` which would need updating to style the label effectively

ciampo commented on packages/components/src/tools-panel/styles.ts:
```diff
@@ -117,6 +119,25 @@ export const ToolsPanelItem = css`
+		}
+	}
+
+	${ BaseControlLabel } {
```
It looks like the `CustomSelectControl`'s label also has a different line height — we may want to also add a temporary patch for that

```suggestion
	.components-custom-select-control__label,
	${ BaseControlLabel } {
```

aaronrobertshaw replied:
Thanks. I've added this along with a comment stating the use of the hardcoded class is temporary.  https://github.com/WordPress/gutenberg/pull/36387/commits/8ae226cf4fd13d155ee8fecafe8dc3e063a68a59

ciampo reviewed (approved):
> This appears to be due to the lack of a standard line-height applied to labels in the Storybook.

Yes, my main point is to say that these fixes to the labels:

- should be applied to the individual components (and not to the `ToolsPanel`)
- should be tested in Storybook first, and in the editor after

I understand the need to apply this fix to the `ToolsPanel` temporarily with this PR, but we should follow-up with an investigation and a refactor of the control components. Ideally, they should all inherit from the same `BaseControl` and share the same label styles.

Finally, before merging, could you also add a quick note to the CHANGELOG? Thank you!

🚀

aaronrobertshaw committed:
Ensure line-height for CustomSelectControl labels

aaronrobertshaw committed:
Add changelog
